### PR TITLE
Add AdmissionController interface

### PR DESCRIPTION
**Pull request by mitalawachat**
Signed-off-by: Mital Awachat <awachatm@amazon.com>

### Description
This pull request serves as a starting point for Admission Control. In this pull request we are adding AdmissionController interface which will be fulfilled by concrete implementation in subsequent pull requests as mentioned in design proposal given below.
 
### Issues Resolved
Feature Proposal: https://github.com/opensearch-project/OpenSearch/issues/1144
Design Proposal: https://github.com/opensearch-project/OpenSearch/issues/3400
 
### Check List
- [-] New functionality includes testing.
  - [-] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/throttle/AdmissionController.java
Comment (nknize) [inline server/src/main/java/org/opensearch/common/throttle/AdmissionController.java:17]: Do we intend for plugins to be able to implement this interface? Can we make this package private? 

If not please mark this with an appropriate opensearch (@opensearch.internal, @opensearch.experimental, @opensearch.api) annotation to document extensibility restrictions.

**Review by nknize — CHANGES_REQUESTED** [review]
This interface needs to be included in a PR that implements it and includes some kind of tests. We don't merge PRs for unused interfaces. If this is a merge of a fairly large downstream implementation make sure it's split into smaller incremental PRs.

Also, we'll want to feature flag this so we can let it bake before releasing GA.

File: server/src/main/java/org/opensearch/common/throttle/AdmissionController.java
Comment (mitalawachat) [reply server/src/main/java/org/opensearch/common/throttle/AdmissionController.java:17]: > This interface needs to be included in a PR that implements it and includes some kind of tests. We don't merge PRs for unused interfaces. If this is a merge of a fairly large downstream implementation make sure it's split into smaller incremental PRs.
> 
> Also, we'll want to feature flag this so we can let it bake before releasing GA.

Thanks for the feedback @nknize 

I planned to raise implementations as part of next pull-requests to keep scope of each pull-request smaller.

I will update this pull request with other suggested changes.

Comment (mitalawachat) [reply server/src/main/java/org/opensearch/common/throttle/AdmissionController.java]: > Do we intend for plugins to be able to implement this interface? Can we make this package private?

Made it package private as it is not intended for plugins to implement this interface.

---

> Also, we'll want to feature flag this so we can let it bake before releasing GA.

Yes, this feature will be behind dynamic setting.

**Review by nknize — CHANGES_REQUESTED** [review]
Thanks for including the requested changes. But we do need to include this interface in a PR that implements it and includes some kind of tests. We don't merge PRs with unused interfaces. If this is a merge of a fairly large downstream implementation make sure it's split into smaller incremental PRs. Can you at least include this with some part of a mechanism that uses it?

**Comment by getsaurabh02** [comment]
+1 on @nknize on having one of the implementations out before we merge just the interface. However, the only difficulty I see is that if we do as a separate followup PR, it will fail checks since interface is not yet available in the core.

@mitalawachat Can we have one of the smaller implementations added to this same PR. That will help validate the consistency and completeness of the Interface being added.

However, if it consists of bigger chunk of changes and we wish to break them up into multiple followup PRs, we should discuss the possibility of a feature-branch as well.

File: server/src/main/java/org/opensearch/common/throttle/AdmissionControlSetting.java
Comment (mitalawachat) [inline server/src/main/java/org/opensearch/common/throttle/AdmissionControlSetting.java:57]: Will fix this along with review comments.

File: server/src/main/java/org/opensearch/common/throttle/AdmissionController.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/common/throttle/AdmissionController.java:65]: Can metric tracking be modelled as a listener

File: server/src/main/java/org/opensearch/common/throttle/RequestSizeAdmissionController.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/common/throttle/RequestSizeAdmissionController.java:60]: do we need a reset? or can it called as a part of the constructor

**Comment by dreamer-89** [comment]
@mitalawachat : Are you still working on this ?

File: server/src/main/java/org/opensearch/common/throttle/AdmissionControlSetting.java
Comment (ketanv3) [inline server/src/main/java/org/opensearch/common/throttle/AdmissionControlSetting.java:18]: We should allow each admission controller to bring their own settings, rather than enforcing this fixed format.

File: server/src/main/java/org/opensearch/common/throttle/AdmissionController.java
Comment (ketanv3) [inline server/src/main/java/org/opensearch/common/throttle/AdmissionController.java:64]: I feel that these interface methods model each admission controller as a token-bucket, which is not always true. For example, an admission controller which only looks at the node's CPU/memory usage doesn't need to acquire/release any tokens.

File: server/src/main/java/org/opensearch/common/throttle/AdmissionController.java
Comment (ketanv3) [inline server/src/main/java/org/opensearch/common/throttle/AdmissionController.java:70]: I'm not sure why this belongs here. We should make admission controllers pluggable.

**Pull request closed without merging**
